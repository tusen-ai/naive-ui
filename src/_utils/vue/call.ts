type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
