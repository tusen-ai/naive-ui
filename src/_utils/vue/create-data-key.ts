export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
