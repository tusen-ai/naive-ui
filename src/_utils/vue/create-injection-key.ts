import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
