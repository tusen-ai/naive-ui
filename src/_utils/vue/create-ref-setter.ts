import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
