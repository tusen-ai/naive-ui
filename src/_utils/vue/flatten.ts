import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
