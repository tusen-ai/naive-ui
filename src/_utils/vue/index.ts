export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
