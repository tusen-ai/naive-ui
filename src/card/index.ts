export { default as NCard, cardProps } from './src/Card'
export type { CardProps, CardSegmented } from './src/Card'
