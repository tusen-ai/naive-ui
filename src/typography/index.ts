export { aProps, default as NA } from './src/a'
export type { AProps } from './src/a'
export { blockquoteProps, default as NBlockquote } from './src/blockquote'
export type { BlockquoteProps } from './src/blockquote'
export {
  headerProps as h1Props,
  headerProps as h2Props,
  headerProps as h3Props,
  headerProps as h4Props,
  headerProps as h5Props,
  headerProps as h6Props
} from './src/create-header'
export type { HeaderProps as H1Props } from './src/create-header'
export type { HeaderProps as H2Props } from './src/create-header'
export type { HeaderProps as H3Props } from './src/create-header'
export type { HeaderProps as H4Props } from './src/create-header'
export type { HeaderProps as H5Props } from './src/create-header'
export type { HeaderProps as H6Props } from './src/create-header'
export { NH1, NH2, NH3, NH4, NH5, NH6 } from './src/headers'
export { default as NHr } from './src/hr'
export { default as NLi } from './src/li'
export { default as NOl, olProps } from './src/ol'
export type { OlProps } from './src/ol'
export { default as NP, pProps } from './src/p'
export type { PProps } from './src/p'
export { default as NText, textProps } from './src/text'
export type { TextProps } from './src/text'
export { default as NUl, ulProps } from './src/ul'
export type { UlProps } from './src/ul'
