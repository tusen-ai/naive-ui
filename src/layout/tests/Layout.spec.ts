import { mount } from '@vue/test-utils'
import { h } from 'vue'
import { NCard } from '../../card'
import { NH2 } from '../../typography'
import {
  NLayout,
  NLayoutContent,
  NLayoutFooter,
  NLayoutHeader,
  NLayoutSider
} from '../index'

describe('n-layout', () => {
  it('should work with import on demand', () => {
    mount(NLayout)
  })

  it('should work with Basic', async () => {
    const wrapper = mount(NLayout, {
      props: {
        'has-sider': true
      },
      slots: {
        default: () => [
          h(NLayoutHeader, null, { default: () => 'test-header' }),
          h(NLayoutContent, null, { default: () => 'test-content' }),
          h(NLayoutSider, null, { default: () => 'test-sider' }),
          h(NLayoutFooter, null, { default: () => 'test-footer' })
        ]
      }
    })

    expect(
      wrapper.find('.n-layout-scroll-container').element.children.length
    ).toBe(4)
    expect(
      wrapper
        .find('.n-layout-scroll-container')
        .element.children[0].getAttribute('class')
    ).toContain('n-layout-header')
    expect(wrapper.find('.n-layout-header').text()).toBe('test-header')
    expect(
      wrapper
        .find('.n-layout-scroll-container')
        .element.children[1].getAttribute('class')
    ).toContain('n-layout-content')
    expect(wrapper.findAll('.n-layout-scroll-container')[1].text()).toBe(
      'test-content'
    )
    expect(
      wrapper
        .find('.n-layout-scroll-container')
        .element.children[2].getAttribute('class')
    ).toContain('n-layout-sider')
    expect(wrapper.find('.n-layout-sider-scroll-container').text()).toBe(
      'test-sider'
    )
    expect(
      wrapper
        .find('.n-layout-scroll-container')
        .element.children[3].getAttribute('class')
    ).toContain('n-layout-footer')
    expect(wrapper.find('.n-layout-footer').text()).toBe('test-footer')
    wrapper.unmount()
  })

  it('should work with `content-style` prop', async () => {
    const wrapper = mount(NLayout, {
      props: {
        'has-sider': true
      },
      slots: {
        default: () =>
          h(
            NLayoutSider,
            { 'content-style': 'padding: 24px' },
            {
              default: () => [
                h(NH2, null, { default: () => 'test1' }),
                h(NH2, null, { default: () => 'test2' })
              ]
            }
          )
      }
    })
    expect(
      wrapper.find('.n-layout-sider-scroll-container').attributes('style')
    ).toContain('padding: 24px')
    wrapper.unmount()
  })

  it('should work with `embedded` prop', async () => {
    const wrapper = mount(NLayout, {
      props: {
        embedded: true
      },
      slots: {
        default: () =>
          h(NCard, null, {
            default: () => 'test'
          })
      }
    })
    expect(wrapper.find('.n-layout').attributes('style')).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should work with `bordered` prop', async () => {
    const wrapper = mount(NLayout, {
      props: {
        'has-sider': true
      },
      slots: {
        default: () => [
          h(
            NLayoutHeader,
            { bordered: true },
            { default: () => 'test-header' }
          ),
          h(NLayoutSider, { bordered: true }, { default: () => 'test-sider' }),
          h(NLayoutFooter, { bordered: true }, { default: () => 'test-footer' })
        ]
      }
    })
    expect(wrapper.find('.n-layout-header').classes()).toContain(
      'n-layout-header--bordered'
    )
    expect(wrapper.find('.n-layout-sider').classes()).toContain(
      'n-layout-sider--bordered'
    )
    expect(wrapper.find('.n-layout-footer').classes()).toContain(
      'n-layout-footer--bordered'
    )
    wrapper.unmount()
  })

  it('should work with `inverted` prop', async () => {
    const wrapper = mount(NLayout, {
      props: {
        'has-sider': true
      },
      slots: {
        default: () => [
          h(
            NLayoutHeader,
            { inverted: true },
            { default: () => 'test-header' }
          ),
          h(NLayoutSider, { inverted: true }, { default: () => 'test-sider' }),
          h(NLayoutFooter, { inverted: true }, { default: () => 'test-footer' })
        ]
      }
    })
    expect(
      wrapper.find('.n-layout-header').attributes('style')
    ).toMatchSnapshot()
    expect(
      wrapper.find('.n-layout-sider').attributes('style')
    ).toMatchSnapshot()
    expect(
      wrapper.find('.n-layout-footer').attributes('style')
    ).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should work with `position` prop', async () => {
    let wrapper = mount(NLayout, {
      slots: {
        default: () => [
          h(NLayoutHeader, null, { default: () => 'test-header' })
        ]
      }
    })
    expect(wrapper.find('.n-layout').classes()).toContain(
      'n-layout--static-positioned'
    )
    wrapper = mount(NLayout, {
      props: {
        position: 'absolute'
      },
      slots: {
        default: () => [
          h(NLayoutHeader, null, { default: () => 'test-header' })
        ]
      }
    })
    expect(wrapper.find('.n-layout').classes()).toContain(
      'n-layout--absolute-positioned'
    )
    wrapper.unmount()
  })

  it('should work with `sider-placement` prop', async () => {
    const wrapper = mount(NLayout, {
      props: {
        hasSider: true
      },
      slots: {
        default: () => [
          h(NLayoutSider, null, { default: () => 'test-sider' }),
          h(NLayoutContent, null, { default: () => 'test-footer' })
        ]
      }
    })
    expect(wrapper.find('.n-layout-sider').classes()).toContain(
      'n-layout-sider--left-placement'
    )

    await wrapper.setProps({ siderPlacement: 'right' })
    expect(wrapper.find('.n-layout-sider').classes()).toContain(
      'n-layout-sider--right-placement'
    )
    wrapper.unmount()
  })
})
