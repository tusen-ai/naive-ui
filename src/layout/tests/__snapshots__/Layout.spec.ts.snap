// Vitest Snapshot v1

exports[`n-layout > should work with \`embedded\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgb(250, 250, 252); --n-text-color: rgb(51, 54, 57);"`;

exports[`n-layout > should work with \`inverted\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgb(0, 20, 40); --n-text-color: #FFF; --n-border-color: rgb(0, 20, 40);"`;

exports[`n-layout > should work with \`inverted\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-toggle-button-color: #FFF; --n-toggle-button-border: 1px solid rgb(239, 239, 245); --n-toggle-bar-color: rgba(191, 191, 191, 1); --n-toggle-bar-color-hover: rgba(153, 153, 153, 1); --n-color: rgb(0, 20, 40); --n-text-color: #FFF; --n-border-color: rgb(0, 20, 40); --n-toggle-button-icon-color: rgb(51, 54, 57); max-width: 272px; width: 272px;"`;

exports[`n-layout > should work with \`inverted\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgb(0, 20, 40); --n-text-color: #FFF; --n-border-color: rgb(0, 20, 40);"`;
